### PR TITLE
refactor: get payment gateways from ERPNext

### PR DESCRIPTION
**Pull request by s-aga-r**
https://github.com/frappe/erpnext/pull/37182

- [x] GoCardless Settings
- [x] GoCardless Mandate
- [x] GoCardless Templates
- [x] Mpesa Settings
- [x] stripe_integration.py
- [x] Dependencies
- [x] Fix Imports
- [x] Fix Unit Tests

### CODE REVIEW

**Comment by ankush** [comment]
Base branch seems to be messed up? Why is it showing these many commits. 


Maybe rebase it on latest develop branch.

**Comment by s-aga-r** [comment]
> Base branch seems to be messed up? Why is it showing these many commits.
> 
> Maybe rebase it on latest develop branch.

FF Rebase won't work, had too many changes in the last 5 years. 

OR we can move without history.

**Comment by s-aga-r** [comment]
@ankush 

Two blockings here:
- `GoCardless Settings` creates `GoCardless Mandate` and `GoCardless Mandate` has a Link field for `ERPNext Customer`. We can create a custom field on installation (?)
- `test_mpesa_settings.py` requires `create_pos_invoice` from ERPNext.

**Comment by blaggacao** [comment]
Maybe `validate_transaction_currency` weirdly called by `erpnext.accounts.doctype.payment_request.payment_request.{get_request_acount,get_payment_url}` and `validate_minimum_transaction_amount` called by `[idem].get_payment_url` should both be folded into `on_payment_request_submission`, instead?

This would reduce the api interface to the necessary minimum and along with https://github.com/frappe/erpnext/pull/37503 would make this refactoring a bit more polished.

However so, at the same time, we need an API expansion in another place (https://github.com/frappe/erpnext/pull/37502) to support known use cases (e.g. Lyra / PayZen) which I'm happy to contribute once stable.

File: payments/payment_gateways/doctype/gocardless_settings/gocardless_settings.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+		else:
+			return "live"
+
+	def validate_transaction_currency(self, currency):
+		if currency not in self.supported_currencies:
+			frappe.throw(
+				_(
+					"Please select another payment method. Go Cardless does not support transactions in currency '{0}'"
+				).format(currency)
+			)
```
Comment (blaggacao) [inline payments/payment_gateways/doctype/gocardless_settings/gocardless_settings.py:99]: Should be folded into `on_payment_request_submission`

File: payments/payment_gateways/doctype/mpesa_settings/mpesa_settings.py [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+class MpesaSettings(Document):
+	supported_currencies = ["KES"]
+
+	def validate_transaction_currency(self, currency):
+		if currency not in self.supported_currencies:
+			frappe.throw(
+				_(
+					"Please select another payment method. Mpesa does not support transactions in currency '{0}'"
+				).format(currency)
+			)
```
Comment (blaggacao) [inline payments/payment_gateways/doctype/mpesa_settings/mpesa_settings.py:29]: Should be folded into `on_payment_request_submission`.

**Review by blaggacao — CHANGES_REQUESTED** [review]
Unless this refactoring is barred from orthogonal improvements / changes. According to above comment, I suggest to:

- Standardize around `on_payment_request_submission` hook for `validate_transaction_currency` & `validate_minimum_transaction_amount` (and all other relevant validations)
- Consequentially, throw on `payment_gateway_validation` instead of casting any errors into silence by default (callers can do so if they so require).
- Amplify the API with `on_payment_authorized_redirect` as described in the linked PR
- Maybe even reduce the interface around `controller.request_for_payment(**payment_record)`

If this refactoring is barred from these changes, in order to facilitate a coordinated implementation, I'd suggest to address these issues before the refactoring is completed in order to avoid having to coordinate an API refactoring across repositories later.

**Comment by s-aga-r** [comment]
@blaggacao do changes separately once this PR is merged.

**Pull request merged**